src/detectors/Rust/RustLanguageDetector.ts

```diff
@@ -0,0 +1,65 @@
+import { ILanguageDetector } from '../ILanguageDetector';
+import { injectable, inject } from 'inversify';
+import { LanguageAtPath, ProgrammingLanguage } from '../../model';
+import { IFileInspector } from '../../inspectors/IFileInspector';
+import { Types } from '../../types';
+import { fileExtensionRegExp, fileNameRegExp, sharedSubpath } from '../utils';
+import lo from 'lodash';
+import * as nodePath from 'path';
+
+@injectable()
+export class RustLanguageDetector implements ILanguageDetector {
+  private fileInspector: IFileInspector;
+
+  constructor(@inject(Types.IFileInspector) fileInspector: IFileInspector) {
+    this.fileInspector = fileInspector;
+  }
+
+  async detectLanguage(): Promise<LanguageAtPath[]> {
+    const result: LanguageAtPath[] = [];
+
+    const manifestFilesPromise = this.fileInspector.scanFor(fileNameRegExp('Cargo.toml'), '/').then((arr) =>
+      arr.map((file) => {
+        return {
+          language: ProgrammingLanguage.Rust,
+          path: nodePath.dirname(file.path),
+        };
+      }),
+    );
+
+    // * `foo/bar.rs` => `foo/`
+    // * `foo/bar/baz.rs, foo/bar/qux.rs` => `foo/bar/`
+    // * `foo/bar.rs, `baz/qux.rs` => `.`
+    // And specially: `foo/src/foo.rs, foo/src/bar.rs` => `foo/` (skips `src`)
+    const rustSourcesPromise = this.fileInspector
+      .scanFor(fileExtensionRegExp(['rs']), '/')
+      .then((files) => lo.uniq(files.map((f) => nodePath.dirname(f.path))))
+      .then((uniqueDirs) => {
+        if (uniqueDirs.length === 0) {
+          return [];
+        }
+
+        let commonPath = sharedSubpath(uniqueDirs);
+        if (nodePath.basename(commonPath) === 'src') {
+          commonPath = nodePath.dirname(commonPath);
+        }
+
+        return [
+          {
+            language: ProgrammingLanguage.Rust,
+            path: commonPath,
+          },
+        ];
+      });
+
+    const manifestFiles = await manifestFilesPromise;
```

Comment:
In an ideal world the promises might be executed on another thread that does IO while the main thread continues execution until it truly needs the result (in this case it prepares all the conditions in promises and then resolves them when it actually needs them).

Whether or not that is actually happening depends on the implementation (i.e. node) and whether the promises are lazy or not, but I believe this pattern _may_ have positive performance impact.

I would also like to come back to your previous questions about `async/await` vs `.then`. When using `await` directly in the top-level function (in this case the `detectLanguage`) the whole function stalls while the value is resolved. That means each step that is now represented by `then` would be a separate top-level await and would prevent the current function from advancing, thus the IO operations to never happen in parallel.

When using `then` (or `await` nested in `then` callbacks) it is actually the internal promise resolution that is stalled, thus allowing the top-level function to continue executing (and preparing other conditions) until it needs the results. This allows (in theory, and depending on how node is interally implemented, maybe even in practice) a performance boost because the IO operations actually run in parallel.

Of course I don't have any benchmarks to back this up so if you'd like me to rewrite it I will, but I am confident in what I said.